core/include/ignition/plugin/Plugin.hh

```diff
@@ -137,6 +137,9 @@ namespace ignition
       public: bool HasInterface(const std::string &_interfaceName,
                                 const bool _demangled = true) const;
 
+      /// \brief Returns the name of the plugin.
+      public: std::string Name() const;
```

Comment:
This function could return by const-reference (i.e. `const std::string&`) instead of by value (i.e. `std::string`) to avoid unnecessary heap allocation and copying.

Reply:
Valid point :)

Reply:
can you please make this change?

Reply:
Please see the comment below. I removed the reference so I can return a value in case 'info' is not initialized.

Reply:
> can you please make this change?

ok, I saw that you did make the change but then reverted it in case the `info` data structure is a `nullptr`. does that work for you @mxgrey?

Reply:
IIRC, if `info` is ever a `nullptr` during a time that a user could call this function, then the `ignition-plugin` library is fundamentally broken somewhere.

But also I don't see where/if that requirement explicitly enforced, so perhaps it's better to not risk returning a reference.

Instead, we could return a `const std::string*` which has a `nullptr` value if `info` is a `nullptr`, or otherwise returns a pointer to the name of the plugin. This would be an equivalent way of avoiding a copy while accommodating the possibility that no `std::string` is actually available.

Reply:
Ok then, I'll use a pointer. (fyi I changed it as a result of the PluginPtr.Empty test crashing)

Reply:
Ah, okay, then I was wrong about `info` always being non-null. I'm glad that test exists :+1:

Reply:
Is the copy really a big concern here? Shouldn't the compiler take care of it with return value optimization (RVO)?

In this case it looks like returning an invalid value is important, so maybe a simple `std::string` isn't expressive enough. I think this is a good use case for `std::optional<std::string>`, which is more expressive than a raw pointer:

```
    std::optional<std::string> Plugin::Name() const
    {
      if (!this->dataPtr->info)
        return std::nullopt;

      return {this->dataPtr->info->name};
    }
```

Reply:
No, RVO doesn't really apply to the copying that we're talking about here.

If a `std::string` is returned by value, then the function requires a copy of the internally stored `name` field to be copied. This copy will be made whether or not the user has any need for a copy to be made. If they only want to view the name, then they have no use for making a copy, and it would be wasting time and resources (albeit, not very significant resources, so this discussion is reaching bike shedding status at this point).

A `const std::string*` is the perfect type expression for what I think this function should be providing, which is this: "Here is an immutable, non-owning view of the plugin's internally-stored name, or a `nullptr` if there is no plugin at all."

I understand that historically raw pointers have had some ambiguity to their meaning, but I think we're far enough into modern C++ that the ambiguity has died out (with a few straggling libraries like Qt). Besides, this meaning for raw pointers is used very consistently throughout the rest of `ignition-plugin`, so I don't see it being an issue here.

Reply:
Ok yeah you're right, RVO doesn't apply here.

I still think that we could leverage modern concepts that C++17 offers instead of potentially ambiguous raw pointers when introducing new API.

Reply:
If we want to use the most correct modern C++ type expression, then we want a [`std::observer_ptr<std::string>`](https://en.cppreference.com/w/cpp/experimental/observer_ptr), but that's unfortunately not released in the Standard yet.

Until `std::observer_ptr` is available, raw pointers are the correct data type for providing a view of an object that might not be available (a reference would be preferable if we can guarantee the availability of the object, but obviously the tests revealed that we can't). Since `ignition-plugin` is a systems programming library, it would be bad practice to force a heap allocation and copy onto a user when they don't need it.

Reply:
> not released in the Standard yet

That's not an option then, right? :upside_down_face: 

> it would be bad practice to force a heap allocation and copy onto a user when they don't need it

Ok, fair enough, I can see how the pointer offers us a way to prevent a copy here. Hopefully we don't run into ownership issues and the caller remembers to make the copy if they need to hold that value beyond the lifetime of info.

Thank you for walking through this with me :bowing_woman: